stl/inc/span

```diff
@@ -0,0 +1,727 @@
+
+    template <_Is_span_compatible_iterator<element_type> _It>
+    constexpr span(_It _First, size_type _Count) noexcept // strengthened
+        : _Mybase(_Count), _Mydata(_STD to_address(_Get_unwrapped_n(_First, _Count))) {
+#if _CONTAINER_DEBUG_LEVEL > 0
+        if constexpr (_Extent != dynamic_extent) {
+            _STL_VERIFY(_Count == _Extent, "span construction from invalid range as count != extent");
+        }
+#endif // _CONTAINER_DEBUG_LEVEL > 0
+    }
+
+    template <_Is_span_compatible_iterator<element_type> _It, _Is_span_compatible_sentinel<_It> _Sentinel>
+    constexpr span(_It _First, _Sentinel _Last) noexcept(noexcept(_Last - _First)) // strengthened
+        : _Mybase(static_cast<size_type>(_Last - _First)), _Mydata(_STD to_address(_First)) {
+        _Adl_verify_range(_First, _Last);
+#if _CONTAINER_DEBUG_LEVEL > 0
+        if constexpr (_Extent != dynamic_extent) {
+            _STL_VERIFY(_Last - _First == _Extent, "span construction from invalid range as last - first != extent");
+        }
+#endif // _CONTAINER_DEBUG_LEVEL > 0
+    }
+
+    template <size_t _Size>
+        requires (_Extent == dynamic_extent || _Extent == _Size)
+    constexpr span(type_identity_t<element_type> (&_Arr)[_Size]) noexcept : _Mybase(_Size), _Mydata(_Arr) {}
+
+    template <size_t _Size>
+        requires (_Extent == dynamic_extent || _Extent == _Size)
+    constexpr span(array<value_type, _Size>& _Arr) noexcept : _Mybase(_Size), _Mydata(_Arr.data()) {}
+
+    template <size_t _Size>
+        requires (_Extent == dynamic_extent || _Extent == _Size) && is_convertible_v<const value_type*, element_type*>
+    constexpr span(const array<value_type, _Size>& _Arr) noexcept : _Mybase(_Size), _Mydata(_Arr.data()) {}
+
+    template <class _Rng>
+        requires (_Extent == dynamic_extent) && _Is_span_compatible_range<_Rng, element_type>
+    constexpr span(_Rng&& _Range)
+        : _Mybase(static_cast<size_type>(_RANGES size(_Range))), _Mydata(_RANGES data(_Range)) {}
+
+    template <class _OtherTy, size_t _OtherExtent>
+        requires (_Extent == dynamic_extent || _Extent == _OtherExtent)
+              && is_convertible_v<_OtherTy (*)[], element_type (*)[]>
+    constexpr span(const span<_OtherTy, _OtherExtent>& _Other) noexcept
+        : _Mybase(_Other.size()), _Mydata(_Other.data()) {}
+    // clang-format on
+
+#else // ^^^ __cpp_lib_concepts / !__cpp_lib_concepts vvv
+
+    template <size_t _Ext = _Extent, enable_if_t<_Ext == 0 || _Ext == dynamic_extent, int> = 0>
+    constexpr span() noexcept {}
+
+    constexpr span(pointer _Ptr, size_type _Count) noexcept // strengthened
+        : _Mybase(_Count), _Mydata(_Ptr) {
+#if _CONTAINER_DEBUG_LEVEL > 0
+        if constexpr (_Extent != dynamic_extent) {
+            _STL_VERIFY(_Count == _Extent, "span construction from invalid range as count != extent");
+        }
+#endif // _CONTAINER_DEBUG_LEVEL > 0
+    }
+
+    constexpr span(pointer _First, pointer _Last) noexcept // strengthened

```

Comment:
Testing has discovered a minor safety issue in the non-concepts fallback - the `(pointer, size_type)` and `(pointer, pointer)` constructors can't reject `span<Base>(Derived*, size_t or Derived*)`. We've decided to do nothing about this (except disable the tests accordingly), since the non-concepts fallback will be fairly short-lived.

Reply:
Another interesting consequence of P1394: `span<int>(some_int_pointer, 0)` works. It was previously ambiguous whether that initialization would call the `(pointer, size)` or `(pointer, pointer)` constructor since `0` is a null pointer constant.

Reply:
Maybe "interesting but largely useless" would be a better description of that fact.